numba/np/arrayobj.py

```diff
@@ -2507,14 +2511,30 @@ def array_record_getattr(context, builder, typ, value, attr):
     newdataptr = cgutils.pointer_add(
         builder, array.data, constoffset,  return_type=rary.data.type,
     )
-    datasize = context.get_abi_sizeof(context.get_data_type(dtype))
-    populate_array(rary,
-                   data=newdataptr,
-                   shape=array.shape,
-                   strides=array.strides,
-                   itemsize=context.get_constant(types.intp, datasize),
-                   meminfo=array.meminfo,
-                   parent=array.parent)
+    if isinstance(dtype, types.NestedArray):
+        datasize = context.get_abi_sizeof(context.get_data_type(dtype.dtype))
+        # new strides = recarray strides + strides of the inner nestedarray
+        strides = cgutils.unpack_tuple(builder, array.strides, typ.ndim)
+        strides += [context.get_constant(types.intp, i) for i in dtype.strides]
+        # new shape = recarray shape + inner dimension from nestedarray
+        shape = cgutils.unpack_tuple(builder, array.shape, typ.ndim)
+        shape += [context.get_constant(types.intp, i) for i in dtype.shape]
+        populate_array(rary,
+                       data=newdataptr,
+                       shape=cgutils.pack_array(builder, shape),
+                       strides=strides,
```

Comment:
It seems inconsistent to `pack_array` `shape` but not `strides`?  API looks like it can take either an iterable python container _or_ an LLVM array, and `shape` and `strides` at this call site are both python lists.
https://github.com/numba/numba/blob/3ec12030586647b0b3b81918256eaa4d11d63780/numba/np/arrayobj.py#L180-L201

Reply:
I think we shouldn't pack the array because `populate_array` will do it for us. Further, I think we needn't have two slightly different calls to `populate_array` - instead we can just determine the correct shape, strides, and datasize for each case in the branches then populate the array after, if we do something like:

```diff
diff --git a/numba/np/arrayobj.py b/numba/np/arrayobj.py
index 7672a2901..97b99d531 100644
--- a/numba/np/arrayobj.py
+++ b/numba/np/arrayobj.py
@@ -2511,30 +2511,29 @@ def array_record_getattr(context, builder, typ, value, attr):
     newdataptr = cgutils.pointer_add(
         builder, array.data, constoffset,  return_type=rary.data.type,
     )
+
     if isinstance(dtype, types.NestedArray):
-        datasize = context.get_abi_sizeof(context.get_data_type(dtype.dtype))
-        # new strides = recarray strides + strides of the inner nestedarray
-        strides = cgutils.unpack_tuple(builder, array.strides, typ.ndim)
-        strides += [context.get_constant(types.intp, i) for i in dtype.strides]
-        # new shape = recarray shape + inner dimension from nestedarray
+        # New shape = recarray shape + inner dimension from nestedarray
         shape = cgutils.unpack_tuple(builder, array.shape, typ.ndim)
         shape += [context.get_constant(types.intp, i) for i in dtype.shape]
-        populate_array(rary,
-                       data=newdataptr,
-                       shape=cgutils.pack_array(builder, shape),
-                       strides=strides,
-                       itemsize=context.get_constant(types.intp, datasize),
-                       meminfo=array.meminfo,
-                       parent=array.parent)
+        # New strides = recarray strides + strides of the inner nestedarray
+        strides = cgutils.unpack_tuple(builder, array.strides, typ.ndim)
+        strides += [context.get_constant(types.intp, i) for i in dtype.strides]
+        # New datasize = size of elements of the nestedarray
+        datasize = context.get_abi_sizeof(context.get_data_type(dtype.dtype))
     else:
+        # New shape, strides, and datasize match the underlying array
+        shape = array.shape
+        strides = array.strides
         datasize = context.get_abi_sizeof(context.get_data_type(dtype))
-        populate_array(rary,
-                       data=newdataptr,
-                       shape=array.shape,
-                       strides=array.strides,
-                       itemsize=context.get_constant(types.intp, datasize),
-                       meminfo=array.meminfo,
-                       parent=array.parent)
+
+    populate_array(rary,
+                   data=newdataptr,
+                   shape=shape,
+                   strides=strides,
+                   itemsize=context.get_constant(types.intp, datasize),
+                   meminfo=array.meminfo,
+                   parent=array.parent)
     res = rary._getvalue()
     return impl_ret_borrowed(context, builder, resty, res)
 
```

Reply:
you're right, I've now changed it. I probably copied that pattern from another function without realizing it was not necessary to do it here.

Reply:
@gmarkall your comment was somehow not showing on screen when I replied to Stuart. so my previous reply was to Stuart's comment. I'll look at yours now.

Reply:
improvement here https://github.com/numba/numba/pull/7359/commits/be6369a07bc83d77de34d1d3425248cfe78c08e0